src/brain/index_selection_job.cpp

```diff
@@ -0,0 +1,174 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// index_selection_job.cpp
+//
+// Identification: src/brain/index_selection_job.cpp
+//
+// Copyright (c) 2015-2018, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "brain/index_selection_util.h"
+#include "brain/index_selection_job.h"
+#include "brain/index_selection.h"
+#include "catalog/query_history_catalog.h"
+#include "catalog/system_catalogs.h"
+#include "optimizer/stats/stats_storage.h"
+
+namespace peloton {
+namespace brain {
+
+#define BRAIN_SUGGESTED_INDEX_MAGIC_STR "brain_suggested_index_"
+
+void IndexSelectionJob::OnJobInvocation(BrainEnvironment *env) {
+  auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+  auto txn = txn_manager.BeginTransaction();
+  LOG_INFO("Started Index Suggestion Task");
+
+  // Generate column stats for all the tables before we begin.
+  // TODO[vamshi]
+  // Instead of collecting stats for every table, collect them only for the
+  // tables
+  // we are analyzing i.e. tables that are referenced in the current workload.
+  optimizer::StatsStorage *stats_storage =
+      optimizer::StatsStorage::GetInstance();
+  ResultType result = stats_storage->AnalyzeStatsForAllTables(txn);
+  if (result != ResultType::SUCCESS) {
+    LOG_ERROR(
+        "Cannot generate stats for table columns. Not performing index "
+        "suggestion...");
+    txn_manager.AbortTransaction(txn);
+    return;
+  }
+
+  // Query the catalog for new SQL queries.
+  // New SQL queries are the queries that were added to the system
+  // after the last_timestamp_
+  auto query_catalog = &catalog::QueryHistoryCatalog::GetInstance(txn);
+  auto query_history =
+      query_catalog->GetQueryStringsAfterTimestamp(last_timestamp_, txn);
+  if (query_history->size() > num_queries_threshold_) {
+    LOG_INFO("Tuning threshold has crossed. Time to tune the DB!");
+
+    // Run the index selection.
+    std::vector<std::string> queries;
+    for (auto query_pair : *query_history) {
+      queries.push_back(query_pair.second);
+    }
+
+    // TODO: Handle multiple databases
+    brain::Workload workload(queries, DEFAULT_DB_NAME, txn);
+    brain::IndexSelection is = {workload, env->GetIndexSelectionKnobs(), txn};
+    brain::IndexConfiguration best_config;
+    is.GetBestIndexes(best_config);
+
+    if (best_config.IsEmpty()) {
+      LOG_INFO("Best config is empty");
+    }
+
+    // Get the existing indexes and drop them.
+    // TODO: Handle multiple databases
+    auto database_object = catalog::Catalog::GetInstance()->GetDatabaseObject(
+      DEFAULT_DB_NAME, txn);
```

Comment:
This is should be addressed at least by having a wrapper function with database name as argument pass in. Multiple database handling is important especially after the catalog refactor.

-- Tianyu, Justin & Tianyi